Fix sdl generator to support Json type

redstab opened the pull request:
Possible fix for #823. There is probably a lot of solutions to the problem.
I am fully open for critique. 

This is first pull request. Tell me if I am doing things wrong.

jtoar commented:
@redstab Worked like a charm for me! Thanks for being so detailed in your issues and pull requests. It really goes a long way. Let's get @cannikin's eyes on this one.

Tobbe commented:
@redstab Thanks for taking this on! I ran in to the broken Json support as well. See here https://community.redwoodjs.com/t/error-unknown-type-json-did-you-mean-json/817

But I also assumed the gql datatype and the prisma datatype should be the same... 🙈 

With your change, does the coercion for Json work on forms? If you scaffold a form from a .sdl.js file with a json field, does it pick up on that and do the `JSON.parse` stuff?

redstab commented:
I have now discovered another issue with the generator. The coercion works as expected but the wrong component gets generated from json scaffold.

Scaffold generates a component `TextArea `which is not a component of my knowledge. 

This results in an react error => 

```
react-dom.development.js:23965 Uncaught Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

Check the render method of `UserExampleForm`.
    at createFiberFromTypeAndProps (react-dom.development.js:23965)
    at createFiberFromElement (react-dom.development.js:23988)
    at createChild (react-dom.development.js:13628)
    at reconcileChildrenArray (react-dom.development.js:13900)
    at reconcileChildFibers (react-dom.development.js:14305)
    at reconcileChildren (react-dom.development.js:16762)
    at updateContextProvider (react-dom.development.js:18256)
    at beginWork (react-dom.development.js:18658)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
    at invokeGuardedCallback (react-dom.development.js:292)
    at beginWork$1 (react-dom.development.js:23203)
    at performUnitOfWork (react-dom.development.js:22154)
    at workLoopSync (react-dom.development.js:22130)
    at performSyncWorkOnRoot (react-dom.development.js:21756)
    at react-dom.development.js:11089
    at unstable_runWithPriority (scheduler.development.js:653)
    at runWithPriority$1 (react-dom.development.js:11039)
    at flushSyncCallbackQueueImpl (react-dom.development.js:11084)
    at flushSyncCallbackQueue (react-dom.development.js:11072)
    at scheduleUpdateOnFiber (react-dom.development.js:21199)
    at Object.enqueueSetState (react-dom.development.js:12639)
    at PageLoader.push.../node_modules/react/cjs/react.development.js.Component.setState (react.development.js:471)
    at _callee$ (page-loader.js:99)
    at tryCatch (runtime.js:45)
    at Generator.invoke [as _invoke] (runtime.js:274)
    at Generator.prototype.<computed> [as next] (runtime.js:97)
    at asyncGeneratorStep (asyncToGenerator.js:5)
    at _next (asyncToGenerator.js:27)
```

**What I found suspicious was that the component name was "TextArea" but not suffixed with Field. When I suffixed the component with "Field" it actually worked!**

So I tracked down the generator to `packages\cli\src\commands\generate\scaffold\scaffold.js` and particularly the json `componentName `in "componentFiles() -> componentMetadata -> Json"

**When I changed the `componentName` to "TextAreaField" it actually generated the correct InputField.** 

I will commit the solution. I also noticed a issue surrounding the display of json. It needs to stringify before rendering. I will also try to fix this

Tobbe commented:
It should be TextAreaField, you are right 👍

redstab commented:
I have now discovered a fix for Showing and Editing json inside the scaffold generator. 

**Original Problem**

Visiting the scaffold url with the list generated from `Names.js.template` (`packages\cli\src\commands\generate\scaffold\templates\components\Names.js.template`) 
returned with a react error similar to this `Objects are not valid as a React child`. 

**Crude Solution** 

Setting `displayFunction` to `JSON.stringify`. This allows to show the objects as strings in the `<td>` tag. But I realized that this will not work with extremly long json objects.

I thought that this was the reason for the `truncate` function for other types. So I needed a `displayFunction` similar to this `((obj) => truncate(JSON.stringify(obj)))`. 

So now I can view json but not edit.

For editing I can not use the `displayFunction` because it now has the truncate function attached to it. Therefore it would cut of the object, which is not that good. 
So I needed som sort of deserilizing function property with value of just `JSON.stringify`. I created a new property `deserilizeFunction` to `componentMetadata` (`packages\cli\src\commands\generate\scaffold\scaffold.js -> componentFiles() -> componentMetadata`) and applied it also to the variable `columns` to access it inside the templates.

Inside the `NameForm.js.template` I updated the `defaultProp` property to `{${column.deserilizeFunction && column.deserilizeFunction + '('}props.${singularCamelName}?.<%= column.name %>${column.deserilizeFunction && ')'}}` because otherwise it would have generated a input (when editing) with the value `[object Object]`. 

Inside the `Name.js.template` I updated the `<td>` tag to `{${column.deserilizeFunction && column.deserilizeFunction + '('}${singularCamelName}.<%= column.name %>${column.deserilizeFunction && ')'}}` because otherwise it would have generated the react error message for children a object. 

So with these changes all the crud operations work with the json type. 

I will commit these changes and will request another review. 

This was my way of solving these problems. Feedback is always welcome.

Tobbe commented:
Maybe you could add a model that uses `Json` to `packages/cli/src/commands/generate/sdl/__tests__/fixtures/schema.prisma` (or update an existing model) in order to provide tests for your change to the sdl generator. And update `packages/cli/src/commands/generate/scaffold/__tests__/fixtures/schema.prisma` to test your changes to the scaffolding

cannikin commented:
Yes, like @Tobbe said I think there are enough changes now that we should get a some tests in place before merging!

Tobbe commented:
For your json display function, how about you add it to `Names.js.template`, like the other display functions? Or was there a specific reason you defined it in `scaffold.js`?

Maybe something like this?

```
const jsonFormat = (obj) => {
  return (
    <pre>
      <code>{JSON.stringify(obj, null, 2)}</code>
    </pre>
  )
}
```

(Just add something to handle really large JSON structures)

Or, perhaps something like this

```
const jsonFormat = (obj) => {
  return truncate(JSON.stringify(obj, null, 2))
}
```

Feel free to totally ignore my suggestions and write whatever you feel would work the best. And please do come up with a better function name, if you can :)

redstab commented:
I have implemented a separate display function for json as @Tobbe suggested.

Tests have also been added for sdl and scaffold json generation as suggested.

This is the first time I have ever written tests, so I would appreciate if you could tell me if the tests are good.

Tobbe commented on packages/cli/src/commands/generate/sdl/__tests__/fixtures/schema.prisma:
```diff
@@ -41,6 +41,12 @@ model Shoe {
 
+model Photo {
+  id       Int    @id @default(autoincrement())
+  name String
```
```suggestion
  name     String
```

Tobbe commented on packages/cli/src/commands/generate/sdl/sdl.js:
```diff
@@ -28,9 +28,15 @@ const modelFieldToSDL = (field, required = true, types = {}) => {
-  return `${field.name}: ${field.isList ? '[' : ''}${field.type}${
-    field.isList ? ']' : ''
-  }${(field.isRequired && required) | field.isList ? '!' : ''}`
+  let dictionary = {
```
```suggestion
  const dictionary = {
```

Tobbe commented:
@redstab Great job on the tests 👍

cannikin commented:
@thedavidprice  any ideas about this error in the Windows build?

```
@redwoodjs/cli: # Fatal error in , line 0
@redwoodjs/cli: # unreachable code
```

Tobbe commented:
About that unreachable code
https://github.com/nodejs/node/issues/34200

Tobbe commented:
@redstab  Now that #840 is merged you'll need to update the code a bit. For example you'll want to use `listDisplayFunction` instead of `displayFunction` for json displayed in the overview/list page. And then you could add another function that displays the json pretty-printed on the show/details page using `displayFunction`

Tobbe commented:
@redstab I love how this looks now! :) Is there anything else you'd like to add or do? Or are you happy with it, and want it merged?

redstab commented:
@Tobbe I feel happy with it, I think its ready to be merged.

Tobbe commented:
Awesome. Then we just have to wait for Rob and/or David to take a final look at it :)

cannikin commented:
Thanks for the work on this!
